File: frontend/src/types/apiKey.ts
// ─────────────────────────────────────────────────────────────────────────────
// API Key Management – TypeScript Type Definitions
// ─────────────────────────────────────────────────────────────────────────────

export interface ApiKey {
  id: string;
  organization_id: string;
  user_id: string;
  name: string;
  /** First 8 characters of the plain key — safe to display */
  prefix: string;
  scopes: string[];
  usage_count: number;
  last_used_at?: string | null;
  expires_at?: string | null;
  is_active: boolean;
  rate_limit_per_minute?: number | null;
  created_at: string;
  updated_at: string;
}

/** Returned once after generation/regeneration — key is shown then hidden */
export interface ApiKeyGenerateResponse {
  key: string;
  api_key: ApiKey;
}

export interface ApiKeyListResponse {
  items: ApiKey[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface ApiKeyCreateRequest {
  name: string;
  expires_at?: string | null;
  scopes?: string[];
  rate_limit_per_minute?: number | null;
}

export interface ApiKeyUpdateRequest {
  name?: string;
  expires_at?: string | null;
  scopes?: string[];
  is_active?: boolean;
  rate_limit_per_minute?: number | null;
}

export interface ApiKeyValidateResponse {
  valid: boolean;
  key_id?: string | null;
  scopes?: string[] | null;
  message: string;
}

export interface ApiKeyListParams {
  search?: string;
  is_active?: boolean;
  scope?: string;
  expires_before?: string;
  expires_after?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  page?: number;
  limit?: number;
}

// ── Static Options ────────────────────────────────────────────────────────────

export const AVAILABLE_SCOPES: { value: string; label: string; description: string }[] = [
  { value: "read:projects",      label: "Read Projects",      description: "View project list and details" },
  { value: "write:projects",     label: "Write Projects",     description: "Create and update projects" },
  { value: "read:api_keys",      label: "Read API Keys",      description: "View API key list and metadata" },
  { value: "write:api_keys",     label: "Write API Keys",     description: "Create, update, delete API keys" },
  { value: "read:organization",  label: "Read Organization",  description: "View organization settings and members" },
  { value: "write:organization", label: "Write Organization", description: "Manage organization settings" },
  { value: "read:audit_logs",    label: "Read Audit Logs",    description: "View all audit log entries" },
  { value: "admin:all",          label: "Admin — All",        description: "Full administrative access" },
];

File: backend/app/schemas/api_key.py
import uuid
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class ApiKeyCreateRequest(BaseModel):
    """Request DTO for creating a new API key."""

    name: str = Field(..., min_length=2, max_length=100, description="Friendly name for the API key")
    expires_at: Optional[datetime] = Field(None, description="Optional expiration datetime (UTC)")
    scopes: List[str] = Field(default_factory=list, description="Permission scopes granted to this key")
    rate_limit_per_minute: Optional[int] = Field(
        None, ge=1, le=10000, description="Max requests per minute (None = unlimited)"
    )


class ApiKeyUpdateRequest(BaseModel):
    """Request DTO for updating an existing API key."""

    name: Optional[str] = Field(None, min_length=2, max_length=100)
    expires_at: Optional[datetime] = None
    scopes: Optional[List[str]] = None
    is_active: Optional[bool] = None
    rate_limit_per_minute: Optional[int] = Field(None, ge=1, le=10000)


class ApiKeyResponse(BaseModel):
    """API key response DTO — never exposes the raw token."""

    id: uuid.UUID
    organization_id: uuid.UUID
    user_id: uuid.UUID
    name: str
    prefix: str
    scopes: List[str]
    usage_count: int
    last_used_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    is_active: bool
    rate_limit_per_minute: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ApiKeyGenerateResponse(BaseModel):
    """Response DTO returned once after key generation or regeneration.
    The plain `key` field is shown exactly once and never stored."""

    key: str = Field(..., description="Plain-text API key — shown once, store it securely")
    api_key: ApiKeyResponse


class ApiKeyListResponse(BaseModel):
    """Paginated list of API keys."""

    items: List[ApiKeyResponse]
    total: int
    page: int
    limit: int
    total_pages: int


class ApiKeyValidateResponse(BaseModel):
    """Lightweight response from key validation endpoint."""

    valid: bool
    key_id: Optional[uuid.UUID] = None
    scopes: Optional[List[str]] = None
    message: str = "OK"
